tools/encryption/pkg/encryption-helper.go

```diff
@@ -0,0 +1,263 @@
+func GenerateEncryptedDocument(htmlStr, accessRequirement string, pubKey *tinkpb.Keyset) (string, error) {
+	km, err := registry.GetKeyManager(aesGcmKeyUrl)
+	if err != nil {
+		return "", err
+	}
+	key, err := generateNewAesGcmKey(km)
+	if err != nil {
+		return "", err
+	}
+	ks := createAesGcmKeyset(key)
+	ksEnc, err := proto.Marshal(&ks)
+	if err != nil {
+		return "", err
+	}
+	kh, err := insecurecleartextkeyset.Read(&keyset.MemReaderWriter{Keyset: &ks})
+	if err != nil {
+		return "", err
+	}
+	parsedHtml, err := html.Parse(strings.NewReader(htmlStr))
+	if err != nil {
+		return "", err
+	}
+	encryptedSections := getAllEncryptedSections(parsedHtml)
+	if err = encryptAllSections(parsedHtml, encryptedSections, kh); err != nil {
+		return "", err
+	}
+	encryptedKey, err := encryptDocumentKey(base64.StdEncoding.EncodeToString(ksEnc), accessRequirement, *pubKey)
+	if err != nil {
+		return "", err
+	}
+	if err = addEncryptedDocumentKeyToHead(encryptedKey, parsedHtml); err != nil {
+		return "", err
+	}
+	return renderNode(parsedHtml, false), nil
+}
+
+// Retrieves Google's public key from the given URL.
+func RetrieveGooglePublicKey(publicKeyUrl string) (tinkpb.Keyset, error) {
+	resp, err := http.Get(publicKeyUrl)
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	r := keyset.NewJSONReader(resp.Body)
+	ks, err := r.Read()
+	if err != nil {
+		return tinkpb.Keyset{}, err
+	}
+	return *ks, nil
+}
+
+// Generates a new AES-GCM key.
+func generateNewAesGcmKey(km registry.KeyManager) ([]byte, error) {
+	p, err := proto.Marshal(&gcmpb.AesGcmKeyFormat{KeySize: aesGcmKeySize})
+	if err != nil {
+		return nil, err
+	}
+	m, err := km.NewKey(p)
+	if err != nil {
+		return nil, err
+	}
+	return proto.Marshal(m)
+}
+
+// Creates an AES-GCM Keyset using the input key.
+func createAesGcmKeyset(key []byte) tinkpb.Keyset {
+	keyData := tinkpb.KeyData{
+		KeyMaterialType: tinkpb.KeyData_SYMMETRIC,
+		TypeUrl:         aesGcmKeyUrl,
+		Value:           key,
+	}
+	keys := []*tinkpb.Keyset_Key{
+		&tinkpb.Keyset_Key{
+			KeyData:          &keyData,
+			Status:           tinkpb.KeyStatusType_ENABLED,
+			KeyId:            1,
+			OutputPrefixType: tinkpb.OutputPrefixType_TINK,
+		},
+	}
+	return tinkpb.Keyset{
+		PrimaryKeyId: 1,
+		Key:          keys,
+	}
+}
+
+// Retrieves all encrypted content sections from the parsed HTML tree.
+func getAllEncryptedSections(parsedHtml *html.Node) []*html.Node {
+	for n := parsedHtml.FirstChild; n != nil; n = n.NextSibling {
+		if (n.DataAtom == atom.Html) && (len(n.Attr) != 0) {
+			for bn := n.FirstChild; bn != nil; bn = bn.NextSibling {
+				if bn.DataAtom == atom.Body {
+					return getEncryptedSectionsDfs(bn)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+// Searches for all <section subscriptions-section="content" encrypted> nodes and returns them.
+func getEncryptedSectionsDfs(bodyNode *html.Node) []*html.Node {
+	var encryptedSections []*html.Node
+	var stack []*html.Node
```

Comment:
This looks more like a "queue" to me, not a stack per se.

Reply:
Switched to using "queue" instead.